[#202] Multi-File QuickOutline View

ghillairet opened the pull request:
See issue #202

Preliminary PR for multi file support in quick outline, should not be merged now, still issues regarding navigation and performance of text filter.

ghillairet committed:
[#202] Multi-File QuickOutline View

tfesenko committed:
Merge branch 'master' into task/202

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/utils/SwaggerFileFinder.java:
```diff
@@ -0,0 +1,128 @@
+public class SwaggerFileFinder {
+
+    /**
+     * Represents the scope for which the JSON reference proposals have to be computed. <br/>
```
`Scope` was moved from `JsonReferenceProposalProvider` and now serves several purposes,  it's not limited to reference proposals anymore. The javadoc should be updated accordingly.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/utils/SwaggerFileFinder.java:
```diff
@@ -0,0 +1,128 @@
+        return visitor.getFiles();
+    }
+
+    public Iterable<IFile> collectFiles(Scope scope, IFile currentFile) {
```
Nice

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/outline/QuickOutline.java:
```diff
@@ -110,6 +135,48 @@ public void modifyText(ModifyEvent e) {
+        if (currentScope == Scope.LOCAL) {
+            treeViewer.setAutoExpandLevel(2);
+        } else {
+            treeViewer.setAutoExpandLevel(0);
```
Cool

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/Model.java:
```diff
@@ -84,6 +97,33 @@ public static Model parseYaml(SwaggerSchema schema, String text) {
+                        .withAttribute(ATTRIBUTE_PARENT, null) //
+                        .withAttribute(ATTRIBUTE_POINTER, JsonPointer.compile("")) //
+                        .withType(AbstractNode.class) //
+                        .readValue(file.getLocationURI().toURL());
```
`ObjectReader` we construct here is identical to the one created in`parseYaml(SwaggerSchema schema, String text)`. Can we reuse the code?

tfesenko reviewed (commented):
Changes overview
Outline:
`QuickOutline` implements new functionality, check `handleMultiView() and`isInvocationEvent`
`OutlineStyledLabelProvider` - defines new labels which may include path.

Related changes:
- `AbstractNode` has a `Model`
- `SwaggerFileFinder` to traverse files, 
  JsonReferenceProposalProvider$FileVisitor, $Scope, #collectFiles() -> SwaggerFileFinder
- a new overload for collectFiles() with a scope as a parameter
- Model now has `path` parameter

ghillairet committed:
[#202] Multi-File QuickOutline View

Fix navigation to elements in non current file from quick outline.
Make sure all models have a path to be display in the quick outline.
Update javadocs.

ghillairet committed:
Merge branch 'master' into task/202

ghillairet committed:
[#202] Multi-File QuickOutline View

Update javadocs.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/NodeDeserializer.java:
```diff
@@ -107,11 +105,10 @@ protected ValueNode deserializeValueNode(JsonParser p, DeserializationContext co
         Object v = context.readValue(p, Object.class);
 
-        ValueNode node = new ValueNode(model, parent, ptr, v);
+        ValueNode node = model.valueNode(parent, ptr, v);
```
I really like the new `objectNode()`, `valueNode` and `arrayNode()` methods. Because they are public we can communicate their intention better by renaming them to something like `addObjectNode()` or `createObjectNode`

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/utils/DocumentUtils.java:
```diff
@@ -104,10 +124,49 @@ public static IFile getWorkspaceFile(IPath path) {
+     * @param path
+     * @param selection
+     */
+    public static void openAndReveal(IPath path, ISelection selection) {
```
So basically we expect that `selection` will wrap an `AbstractNode`, right?
